File: clojure/triangle/src/triangle.clj
(ns triangle)

(defn is-valid? [x y z]
  (and (> (+ x y) z) (> (+ x z) y) (> (+ y z) x)))

(defn equilateral? [x y z]
  (and (is-valid? x y z) (= x y z)))

(defn isosceles? [x y z]
  (and (is-valid? x y z) (or (= x y) (= x z) (= y z))))

(defn scalene? [x y z]
  (and (is-valid? x y z) (not= x y) (not= y z)))

File: r/triangle/triangle.R
triangle <- function(x, y, z) {
  stopifnot(x > 0 && y > 0 && z > 0)
  stopifnot(x + y >= z, x + z >= y, y + z >= x)

  sides = c(x, y, z)
  num_unique_sides = length(unique(sides))
  if (num_unique_sides == 1) {
    structure(sides, class = c("equilateral", "isosceles"))
  } else if (num_unique_sides == 2) {
    structure(sides, class = "isosceles")
  } else {
    structure(sides, class = "scalene")
  }
}
